tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Christoph Läubrich and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.tycho.core.maven;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.maven.SessionScoped;
+import org.apache.maven.artifact.ArtifactUtils;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.Model;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.repository.internal.MavenWorkspaceReader;
+import org.codehaus.plexus.component.annotations.Component;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.repository.WorkspaceReader;
+import org.eclipse.aether.repository.WorkspaceRepository;
+import org.eclipse.tycho.PackagingType;
+
+/**
+ * An implementation of a workspace reader similar to what maven does but for the Tycho packaging
+ * types
+ */
+@SessionScoped
+@Component(role = WorkspaceReader.class, hint = TychoReactorReader.HINT)
+public class TychoReactorReader implements MavenWorkspaceReader {
+    static final String HINT = "tycho-reactor";
+
+    private final Map<String, MavenProject> projectsByGAV;
+    private final WorkspaceRepository repository;
+
+    @Inject
+    public TychoReactorReader(MavenSession session) {
+        this.projectsByGAV = session.getProjects().stream()
+                .collect(toMap(s -> ArtifactUtils.key(s.getGroupId(), s.getArtifactId(), s.getVersion()), identity()));
+        repository = new WorkspaceRepository(HINT, null);
+    }
+
+    @Override
+    public WorkspaceRepository getRepository() {
+        return repository;
+    }
+
+    @Override
+    public File findArtifact(Artifact artifact) {
+        return getTychoReactorProject(artifact).map(project -> {
+            org.apache.maven.artifact.Artifact mainArtifact = project.getArtifact();
+            if (mainArtifact != null && mainArtifact.getFile() != null && mainArtifact.getFile().exists()) {
+                return mainArtifact.getFile();
+            }
+            return project.getBasedir();
```

Comment:
Not sure if it's more correct to use `project.getBasedir()` or `project.getBuild().getOutputDirectory()` (latter is what I currently return).

`ReactorReader` does this:
```java
if (isTestArtifact( artifact )) {
    if (project.hasLifecyclePhase("test-compile")) {
        return new File(project.getBuild().getTestOutputDirectory());
    }
} else {
    String type = artifact.getProperty("type", "");
    if (project.hasLifecyclePhase("compile") && COMPILE_PHASE_TYPES.contains(type)) {
        return new File(project.getBuild().getOutputDirectory());
    }
}
```
And that's why I've picked `project.getBuild().getOutputDirectory()`.

Reply:
Tycho uses the basedir in such situations already so this is consistent (even though a bit bad and we might later want to get rid of it).